lib/json/pure/generator.rb

```diff
@@ -286,12 +288,70 @@ def to_h
         # created this method raises a
         # GeneratorError exception.
         def generate(obj)
-          result = obj.to_json(self)
+          if @indent.empty? and @space.empty? and @space_before.empty? and @object_nl.empty? and @array_nl.empty? and
+              !@ascii_only and !@script_safe and @max_nesting == 0 and !@strict
+            result = generate_json(obj, '')
+          else
+            result = obj.to_json(self)
+          end
           JSON.valid_utf8?(result) or raise GeneratorError,
             "source sequence #{result.inspect} is illegal/malformed utf-8"
           result
         end
 
+        # Handles @allow_nan, @buffer_initial_length, other ivars must be the default value (see above)
+        private def generate_json(obj, buf)
+          case obj
+          when Hash
+            buf << '{'.freeze
+            first = true
```

Comment:
Instead of maintaining the extra variable for the state (`first`). We can use `each_with_index`
```ruby
  when Hash
    buf << '{'
    obj.each_with_index do |(k, v), i|
      buf << ',' unless i.zero?
      fast_serialize_string(k.to_s, buf)
      buf << ':'
      generate_json(v, buf)
    end
    buf << '}'
```

Reply:
That's likely much slower as it causes more Array allocations.
Also the code is more similar to Hash#to_json this way.